src/backends/filesystem.ts

```diff
@@ -0,0 +1,619 @@
+/**
+ * FilesystemBackend: Read and write files directly from the filesystem.
+ *
+ * Security and search upgrades:
+ * - Secure path resolution with root containment when in virtual_mode (sandboxed to cwd)
+ * - Prevent symlink-following on file I/O using O_NOFOLLOW when available
+ * - Ripgrep-powered grep with JSON parsing, plus regex fallback
+ *   and optional glob include filtering, while preserving virtual path behavior
+ */
+
+import * as fs from "fs/promises";
+import * as fsSync from "fs";
+import * as path from "path";
+import { spawn } from "child_process";
+import fg from "fast-glob";
+import micromatch from "micromatch";
+import type {
+  BackendProtocol,
+  EditResult,
+  FileInfo,
+  GrepMatch,
+  WriteResult,
+} from "./protocol.js";
+import {
+  checkEmptyContent,
+  formatContentWithLineNumbers,
+  performStringReplacement,
+} from "./utils.js";
+
+/**
+ * Backend that reads and writes files directly from the filesystem.
+ *
+ * Files are accessed using their actual filesystem paths. Relative paths are
+ * resolved relative to the current working directory. Content is read/written
+ * as plain text, and metadata (timestamps) are derived from filesystem stats.
+ */
+export class FilesystemBackend implements BackendProtocol {
+  private cwd: string;
+  private virtualMode: boolean;
+  private maxFileSizeBytes: number;
+
+  constructor(
+    options: {
+      rootDir?: string;
+      virtualMode?: boolean;
+      maxFileSizeMb?: number;
+    } = {},
+  ) {
+    const { rootDir, virtualMode = false, maxFileSizeMb = 10 } = options;
+    this.cwd = rootDir ? path.resolve(rootDir) : process.cwd();
+    this.virtualMode = virtualMode;
+    this.maxFileSizeBytes = maxFileSizeMb * 1024 * 1024;
+  }
+
+  /**
+   * Resolve a file path with security checks.
+   *
+   * When virtualMode=true, treat incoming paths as virtual absolute paths under
+   * this.cwd, disallow traversal (.., ~) and ensure resolved path stays within root.
+   * When virtualMode=false, preserve legacy behavior: absolute paths are allowed
+   * as-is; relative paths resolve under cwd.
+   *
+   * @param key - File path (absolute, relative, or virtual when virtualMode=true)
+   * @returns Resolved absolute path string
+   * @throws Error if path traversal detected or path outside root
+   */
+  private resolvePath(key: string): string {
+    if (this.virtualMode) {
+      const vpath = key.startsWith("/") ? key : "/" + key;
+      if (vpath.includes("..") || vpath.startsWith("~")) {
+        throw new Error("Path traversal not allowed");
+      }
+      const full = path.resolve(this.cwd, vpath.substring(1));
+      const relative = path.relative(this.cwd, full);
+      if (relative.startsWith("..") || path.isAbsolute(relative)) {
+        throw new Error(`Path: ${full} outside root directory: ${this.cwd}`);
+      }
+      return full;
+    }
+
+    if (path.isAbsolute(key)) {
+      return key;
+    }
+    return path.resolve(this.cwd, key);
```

Comment:
I think even with `virtualMode: false` we do want to avoid resolving outside of `cwd`?